hacspec/src/machine_integers.rs

```diff
@@ -286,90 +323,114 @@ macro_rules! implement_secret_signed_mi {
-            const NUM_BITS: u32 = $bits;
-            const ZERO: Self = $t(0);
-            const ONE: Self = $t(1);
-            const TWO: Self = $t(2);
+            const NUM_BITS: usize = $bits;
+
+            #[inline]
+            #[cfg_attr(feature = "use_attributes", library(hacspec))]
+            fn ZERO() -> Self {
+                $t(0)
+            }
+            #[inline]
+            #[cfg_attr(feature = "use_attributes", library(hacspec))]
+            fn ONE() -> Self {
+                $t(1)
+            }
+            #[inline]
+            #[cfg_attr(feature = "use_attributes", library(hacspec))]
+            fn TWO() -> Self {
+                $t(2)
+            }
 
             #[inline]
-            #[cfg_attr(feature="use_attributes", library(hacspec))]
+            #[cfg_attr(feature = "use_attributes", library(hacspec))]
             fn from_literal(val: u128) -> Self {
                 Self::classify(val as $base)
             }
+
+            #[inline]
+            #[cfg_attr(feature = "use_attributes", library(hacspec))]
+            fn from_hex_string(s: &String) -> Self {
+                Self::classify(<$base>::from_str_radix(s.trim_start_matches("0x"), 16).unwrap())
+            }
         }
-        impl NumericBase for $t {
+        impl Numeric for $t {
             /// Return largest value that can be represented.
-            #[cfg_attr(feature="use_attributes", library(hacspec))]
+            #[cfg_attr(feature = "use_attributes", library(hacspec))]
             fn max_val() -> Self {
                 Self::from(<$base>::max_value())
             }
 
-            #[cfg_attr(feature="use_attributes", library(hacspec))]
+            #[cfg_attr(feature = "use_attributes", library(hacspec))]
             fn wrap_add(self, rhs: Self) -> Self {
                 self + rhs
             }
-            #[cfg_attr(feature="use_attributes", library(hacspec))]
+            #[cfg_attr(feature = "use_attributes", library(hacspec))]
             fn wrap_sub(self, rhs: Self) -> Self {
                 self - rhs
             }
-            #[cfg_attr(feature="use_attributes", library(hacspec))]
+            #[cfg_attr(feature = "use_attributes", library(hacspec))]
             fn wrap_mul(self, rhs: Self) -> Self {
                 self * rhs
             }
-            #[cfg_attr(feature="use_attributes", library(hacspec))]
+            #[cfg_attr(feature = "use_attributes", library(hacspec))]
             fn wrap_div(self, _rhs: Self) -> Self {
                 unimplemented!();
             }
 
             /// `self ^ exp` where `exp` is a `u32`.
-            /// **Note:** the exponent `exp` MUST not be secret.
-            #[cfg_attr(feature="use_attributes", library(hacspec))]
-            fn pow(self, exp: u32) -> Self {
+            /// **Note:** the exponent `exp` MUST NOT be secret.
+            #[cfg_attr(feature = "use_attributes", library(hacspec))]
+            fn exp(self, exp: u32) -> Self {
```

Comment:
Should we type the exponent as a `usize` like rotation values ?

Reply:
Hm, yeah maybe. I switched to `usize` for rotation and shift because that's what some builtin function needed (can't recall which one :grimacing: ).